### PR TITLE
ACTIN-1831: Add external configuration of panel gene lists 

### PR DESCRIPTION
**Pull request by pauldwolfe**
Replace the hardcoded list of genes with a TSV file configuration.

The TSV file will have two column, a name (regex) and a gene. The
name unfortunately needs to be duplicated for each gene but that's
just the limitation of using TSV.

### CODE REVIEW

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    companion object {
+        fun create(panelGeneListTsvPath: String): PanelGeneList {
+
+            class PanelGeneEntry(val testNameRegex: String, val gene: String)
+
+            val entries = CsvMapper().apply { registerModule(KotlinModule.Builder().build()) }.readerFor(PanelGeneEntry::class.java)
+                .with(CsvSchema.emptySchema().withHeader().withColumnSeparator('\t')).readValues<PanelGeneEntry>(File(panelGeneListTsvPath))
+            return PanelGeneList(entries.readAll().groupBy { it.testNameRegex }.mapKeys { Regex(it.key) }
+                .mapValues { it.value.map { e -> e.gene } })
```
Comment (jbartletthmf) [inline clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt:25]: You could pass in a value selector to `groupBy` here:
```suggestion
            return PanelGeneList(entries.readAll().groupBy(PanelGeneEntry::testNameRegex, PanelGeneEntry::gene).mapKeys { Regex(it.key) })
```

File: clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneListTest.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+class PanelGeneListTest {
+    
+    @Test
+    fun `Should read from panel gene list TSV and match gene lists on test name regex`() {
+        val geneList = PanelGeneList.create(ResourceLocator.resourceOnClasspath("panel_gene_list/panel_gene_list.tsv"))
+        val oncoPanel = geneList["HMF oncopanel "]
+        assertThat(oncoPanel).containsExactly("ABCB1", "ABL1")
+        val avlPanelGenes = geneList["an ngs test"]
+        assertThat(avlPanelGenes).containsExactly("EGFR")
```
Comment (nedleitch) [inline clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneListTest.kt:15]: Might be interesting to put something here that makes sure we don't get unexpected results if the input has a pipe character, I'm sure it's working but maybe worth doing anyway?

Also, the English words "leavings" and "angst" would both match this regex I think - would it be worth padding (as in making sure `ngs` only appears as a token unto itself) or anchoring for that one case?

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import com.fasterxml.jackson.dataformat.csv.CsvSchema
+import com.fasterxml.jackson.module.kotlin.KotlinModule
+import java.io.File
+
+class PanelGeneList(private val panelRegexToGenes: Map<Regex, List<String>>) {
+
+    operator fun get(input: String): Set<String> {
+        return panelRegexToGenes
+            .filter { (regex, _) -> regex.containsMatchIn(input.lowercase()) }
+            .flatMap { (_, value) -> value }
```
Comment (jbartletthmf) [inline clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt:13]: or `.flatMap { it.value }`, or `.values.flatten()`.

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package com.hartwig.actin.clinical.feed.standard
+
+import com.fasterxml.jackson.dataformat.csv.CsvMapper
+import com.fasterxml.jackson.dataformat.csv.CsvSchema
+import com.fasterxml.jackson.module.kotlin.KotlinModule
+import java.io.File
+
+class PanelGeneList(private val panelRegexToGenes: Map<Regex, List<String>>) {
+
+    operator fun get(input: String): Set<String> {
```
Comment (jbartletthmf) [inline clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt:10]: The operator fun is cool here, but I find it a little surprising that what looks like a map lookup is actually a regex evaluation than returns a non-null set for every query. Would it be better to name this function descriptively like `listGenesForPanel`?

File: clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneListTest.kt (hunk 2 above)
Comment (jbartletthmf) [reply clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneListTest.kt:15]: Good questions. Another option to increase specificity would be to make the regex matching in the code case-sensitive, and then write case-insensitive regexes where appropriate. That might help if we need to match "NGS" but not "rings".

Comment (pauldwolfe) [reply clinical/src/test/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneListTest.kt:15]: This is just a test, you can put any regex you want into the file and anchor as appropriate. In practice these aren't really relevant edge cases.

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt (hunk 3 above)
Comment (pauldwolfe) [reply clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt:13]: yup, AI generated :)

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt (hunk 4 above)
Comment (pauldwolfe) [reply clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt:10]: you're no fun @jbartletthmf ;)

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt (hunk 3 above)
Comment (jbartletthmf) [reply clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt]: Eventually I'll train the LLMs to anticipate my PR feedback 🙃

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt (hunk 4 above)
Comment (jbartletthmf) [reply clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt]: The best kind of fun is the self-documenting kind 🤓

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package com.hartwig.actin.clinical.feed.standard
+
+import com.fasterxml.jackson.dataformat.csv.CsvMapper
+import com.fasterxml.jackson.dataformat.csv.CsvSchema
+import com.fasterxml.jackson.module.kotlin.KotlinModule
+import java.io.File
+
+class PanelGeneList(private val panelRegexToGenes: Map<Regex, List<String>>) {
+
+    fun matchGenesForTest(testName: String): Set<String> {
```
Comment (jbartletthmf) [inline clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/standard/PanelGeneList.kt:10]: I don't love the naming still. Why does the function refer to "Test" while the class refers to "Panel"? In general, I think it's good to minimize our use of "test" to describe anything that doesn't pertain to our test suites. I also think that `match` doesn't obviously convey what's being returned here compared to `find`, `list`, `filter`, `provide`, etc.

**Review by jbartletthmf — APPROVED** [review]
I still have some thoughts on naming, but it's good to go.
